### ISSUE TITLE
Entity-Relationship Model: New relationship types not appearing and lack of bidirectional visibility

### ISSUE THREAD

**Issue by aseyedia** [DSpace/DSpace#9729]
## Describe the bug
I'm trying to set up our Entity-Relationship model such that it resembles the following:

![image](https://github.com/user-attachments/assets/85ad1998-1359-4a13-b532-448b8d57352d)

We would like two-way relationships between all linked Entities inside of our PEDSpace Entity world, and we would like to be able to see all linked Entities per Item page.

In order to do that, I have customized the following:

1. `relationship-types.xml`
2. `virtual-metadata.xml`
3. `relationship-formats.xml` - though I am not sure if that's relevant.

The issue that I am experiencing is twofold:

### 1. Entities Are not Being Updated with New Relationship Types

I had originally generated and applied an initial set of Entity-Relationships in `relationship-types.xml` that did not have the most descriptive relationship names. For example, these relationship types had names such as `informsPhenotype` when a Concept Set would inform a Phenotype (and vice versa `isInformedByConceptSet`). I thought those names were too ambiguous, and, inspired by the default DSpace relationship type names, I created a new set of relationship types that followed the format:

```
is[Entity A]Of[Entity B]
```

Such that `isInformedByConceptSet` is now `isPhenotypeOfConceptSet`. However, these relationship types are not showing up under the `Edit > Relationships` page of a given typed DSpace Item:

![image](https://github.com/user-attachments/assets/39588a30-e736-442d-9d95-b3a27e387560)

Only the old DSpace Entity types are showing up here, not the new ones. I have updated the `relationship-types.xml` file appropriately and can confirm that they're in the backend:

![image](https://github.com/user-attachments/assets/3802b47a-b11c-4ca5-b8ce-bc03568bee83)

But for some reason, only the old ones work. I have even deleted the old `relationship_types` rows and reassigned the actually relationships in `relationship` to the desired new `relationship_types` row and that yielded no results. So, for this example, all of our relationships in `relationship` right now are only `id` `18`: `resultsInConceptSet/isResultOfStudy`. However, I want them to be `id` `32`: `isStudyOfConceptSet/isConceptSetOfStudy`.

![image](https://github.com/user-attachments/assets/578c1f20-adfb-4ac7-8c7e-c002852c98aa)

I deleted all of the old rows and reassigned all `type_id`s from `relationship` from their previous value to `32`. That did nothing about my problem; the relationship fields in the Item edit page just disappeared, and nothing replaced them. So I reinserted the old rows back into `relationship_types` and reassigned the `type_id`s from `relationship` back to `18` and the functionality was restored.

Furthermore, each item has these `relationship.is*Of.[Entity]` fields that are totally inexplicable. I can sense they are the result of some incorrect or incomplete configuration, but I'm not entirely sure what:

![image](https://github.com/user-attachments/assets/49d7062d-9327-4a4a-9244-9a76709fe8e4)

They must be using the old `relationship_types`, but what's strange is that when I had deleted the old relationship types, these "ghost" relationships were still there.

## 2. Entities do not Exhibit Expected Entity-Relationship Reciprocity

![image](https://github.com/user-attachments/assets/7de88b78-6b70-44ac-80e2-9266626b9aa5)

The fact that I am able to see these linked Items, not only on the edit page but also on the Simple Item view of a Study, is huge. However, I would like for Item which is linked to these Items to **also** be visible on the linked Item page. For example, for the Study "Developing Computable Phenotype for Autism," I would like to link to Concept Set "Antipsychotics," which should have a `isResultOfStudy` (using the old relationship types). However, when I go to that page (for which I have configured a custom Simple Item view by copying `publication.component.*` and adapting it for the Concept Set Entity type, **and** importing that new `conceptset.component.ts` into `eager-theme.module.ts`), I am unable to see the linked Study, and when I go to the Edit page, I see this:

![image](https://github.com/user-attachments/assets/6110526a-9db5-4bf1-9f19-2913eceb3632)

It's showing up under `relationships.isResultOf.Study`. All of the other relationship types are inexplicable to me:

![image](https://github.com/user-attachments/assets/99710e53-271f-462b-a515-2f644889e3e1)

## TL;DR

What I think the fundamental problem is, and what I am struggling to wrap my mind around, is this: why won't DSpace use the new, more complete set of Entity-Relationship definitions, which I have updated in the aforementioned `xml` files and initialized with `initialize-entities`? Where is DSpace choosing which `relationship_types` each Entity Item uses?

PS - this is a dev instance so I'm not overly-concerned about uptime or how it looks to users right now.

## To Reproduce
1. Customize `relationship-types.xml` and call `./bin/dspace initialize-entities -f /path/to/relationship-types.xml`.
4. Change `relationship-types.xml`. 
5. Customize `virtual-metadata.xml`.
6. Call `./bin/dspace initialize-entities -f /path/to/relationship-types.xml` again and reboot DSpace to register `virtual-metadata-xml`.

## Expected behavior

I want DSpace to use the new relationship types configuration and not use the old ones. I have attached the aforementioned `xml` files to this issue:

- [relationship-types.xml](https://gist.github.com/aseyedia/c7457323fd7964009b7aeeaa970ccf5b#file-relationship-types-xml)
- [virtual-metadata.xml](https://gist.github.com/aseyedia/c7457323fd7964009b7aeeaa970ccf5b#file-virtual-metadata-xml)
- [relationship-formats.xml](https://gist.github.com/aseyedia/c7457323fd7964009b7aeeaa970ccf5b#file-relationship-formats-xml)

## Related work
- [DSpace.org Slack Post](https://dspace-org.slack.com/archives/C3V628QNN/p1721839791910189)
- [DSpace Google Groups Tech Support Post](https://groups.google.com/g/dspace-tech/c/EPlZ3A5zT2s/m/aidfhWZoBgAJ)

**Comment by aseyedia** [comment]
```xml
    <!-- Used only to READ "submitter" facets (managed programmatically at SolrServiceImpl) -->
    <bean id="searchFilterSubmitter"
        class="org.dspace.discovery.configuration.DiscoverySearchFilterFacet">
        <property name="indexFieldName" value="submitter" />
        <property name="type" value="authority" />
        <property name="metadataFields">
            <list>
                <value>placeholder.placeholder.placeholder</value>
            </list>
        </property>
    </bean>

    <bean id="searchFilterIsOrgUnitOfProjectRelation" class="org.dspace.discovery.configuration.DiscoverySearchFilter">

        <property name="indexFieldName" value="isOrgUnitOfProject"/>
        <property name="metadataFields">
            <list>
                <value>relation.isOrgUnitOfProject.latestForDiscovery</value>
            </list>
        </property>
        <property name="isOpenByDefault" value="false"/>
        <property name="pageSize" value="10"/>
    </bean>
```

I am getting a strong feeling that this might have something to do with `discovery.xml`.

**Comment by aseyedia** [comment]
It also occurred to me that I might have to reindex DSpace and I got this cascading error message:

```sh
[dspace@pedsdspace01 dspace]$ ./bin/dspace index-discovery -b
The script has started
(Re)building index from scratch.
org.apache.solr.client.solrj.impl.HttpSolrClient$RemoteSolrException: Error from server at http://localhost:8983/solr/search: ERROR: [doc=Item-d545a793-c766-450c-880a-7b073a9639ae] Error adding field 'dspace.entity.type'='[ConceptSet, ConceptSet]' msg=Multiple values encountered for non multiValued copy field search.entitytype: ConceptSet
	at org.apache.solr.client.solrj.impl.HttpSolrClient.executeMethod(HttpSolrClient.java:681)
	at org.apache.solr.client.solrj.impl.HttpSolrClient.request(HttpSolrClient.java:266)
	at org.apache.solr.client.solrj.impl.HttpSolrClient.request(HttpSolrClient.java:248)
	at org.apache.solr.client.solrj.SolrRequest.process(SolrRequest.java:225)
	at org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:177)
	at org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:138)
	at org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:156)
	at org.dspace.discovery.indexobject.IndexFactoryImpl.writeDocument(IndexFactoryImpl.java:153)
	at org.dspace.discovery.indexobject.ItemIndexFactoryImpl.writeDocument(ItemIndexFactoryImpl.java:635)
	at org.dspace.discovery.indexobject.ItemIndexFactoryImpl.writeDocument(ItemIndexFactoryImpl.java:81)
	at org.dspace.discovery.SolrServiceImpl.update(SolrServiceImpl.java:170)
	at org.dspace.discovery.SolrServiceImpl.indexContent(SolrServiceImpl.java:159)
	at org.dspace.discovery.SolrServiceImpl.updateIndex(SolrServiceImpl.java:349)
	at org.dspace.discovery.SolrServiceImpl.updateIndex(SolrServiceImpl.java:336)
	at org.dspace.discovery.SolrServiceImpl.createIndex(SolrServiceImpl.java:304)
	at org.dspace.discovery.IndexClient.internalRun(IndexClient.java:138)
	at org.dspace.scripts.DSpaceRunnable.run(DSpaceRunnable.java:150)
	at org.dspace.app.launcher.ScriptLauncher.executeScript(ScriptLauncher.java:154)
	at org.dspace.app.launcher.ScriptLauncher.handleScript(ScriptLauncher.java:132)
	at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:99)
```

**Comment by aseyedia** [comment]
Now I'm finding, maybe separate from this issue, that there is something wrong with my DSpace instance:

```
[dspace@pedsdspace01 dspace]$ ./bin/dspace healthcheck

#### General Information [took: 0s] [# lines: 5]
Generated: Wed Jul 31 16:39:30 EDT 2024
From - Till: 2024-07-24 - 2024-07-31
Url: https://pedsdspace01.research.chop.edu

Assetstore size     : 730 MB
Log dir size        : 799 MB

###############################


#### Checksum [took: 9s] [# lines: 1]
No md5 checks made!

###############################


#### Item summary [took: 0s] [# lines: 51]
Exception occurred when processing report - org.hibernate.exception.SQLGrammarException: JDBC exception executing SQL [select c1_1.uuid,c1_2.uuid,c1_1.admin,c1_1.collection_id,c1_1.logo_bitstream_id,c1_1.submitter,c1_1.template_item_id,sum(b2_1.size_bytes) from public.item i1_0 join public.collection2item c1_0 on i1_0.uuid=c1_0.item_id join (public.collection c1_1 join public.dspaceobject c1_2 on c1_1.uuid=c1_2.uuid) on c1_1.uuid=c1_0.collection_id join public.item2bundle b1_0 on i1_0.uuid=b1_0.item_id join public.bundle b1_1 on b1_1.uuid=b1_0.bundle_id join public.bundle2bitstream b2_0 on b1_1.uuid=b2_0.bundle_id join public.bitstream b2_1 on b2_1.uuid=b2_0.bitstream_id group by c1_1.uuid,case when c1_1.uuid is not null then 3 end,c1_1.admin,c1_1.collection_id,c1_1.logo_bitstream_id,c1_1.submitter,c1_1.template_item_id] [ERROR: column "c1_2.uuid" must appear in the GROUP BY clause or be used in an aggregate function
  Position: 18] [n/a]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:91)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:58)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:94)
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:264)
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.getResultSet(DeferredResultSetAccess.java:167)
	at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.advanceNext(JdbcValuesResultSetImpl.java:218)
	at org.hibernate.sql.results.jdbc.internal.JdbcValuesResultSetImpl.processNext(JdbcValuesResultSetImpl.java:98)
	at org.hibernate.sql.results.jdbc.internal.AbstractJdbcValues.next(AbstractJdbcValues.java:19)
	at org.hibernate.sql.results.internal.RowProcessingStateStandardImpl.next(RowProcessingStateStandardImpl.java:66)
	at org.hibernate.sql.results.spi.ListResultsConsumer.consume(ListResultsConsumer.java:202)
	at org.hibernate.sql.results.spi.ListResultsConsumer.consume(ListResultsConsumer.java:33)
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.doExecuteQuery(JdbcSelectExecutorStandardImpl.java:209)
	at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.executeQuery(JdbcSelectExecutorStandardImpl.java:83)
	at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:76)
	at org.hibernate.sql.exec.spi.JdbcSelectExecutor.list(JdbcSelectExecutor.java:65)
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.lambda$new$2(ConcreteSqmSelectQueryPlan.java:137)
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:381)
	at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performList(ConcreteSqmSelectQueryPlan.java:303)
	at org.hibernate.query.sqm.internal.QuerySqmImpl.doList(QuerySqmImpl.java:509)
	at org.hibernate.query.spi.AbstractSelectionQuery.list(AbstractSelectionQuery.java:427)
	at org.hibernate.query.Query.getResultList(Query.java:120)
	at org.dspace.content.dao.impl.CollectionDAOImpl.getCollectionsWithBitstreamSizesTotal(CollectionDAOImpl.java:179)
	at org.dspace.content.CollectionServiceImpl.getCollectionsWithBitstreamSizesTotal(CollectionServiceImpl.java:925)
	at org.dspace.health.ItemCheck.getCollectionSizesInfo(ItemCheck.java:144)
	at org.dspace.health.ItemCheck.run(ItemCheck.java:75)
	at org.dspace.health.Check.report(Check.java:31)
	at org.dspace.health.Report.run(Report.java:69)
	at org.dspace.health.Report.main(Report.java:196)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:283)
	at org.dspace.app.launcher.ScriptLauncher.handleScript(ScriptLauncher.java:134)
	at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:99)
Caused by: org.postgresql.util.PSQLException: ERROR: column "c1_2.uuid" must appear in the GROUP BY clause or be used in an aggregate function
  Position: 18
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194)
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:137)
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeQuery(DelegatingPreparedStatement.java:123)
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeQuery(DelegatingPreparedStatement.java:123)
	at org.hibernate.sql.results.jdbc.internal.DeferredResultSetAccess.executeQuery(DeferredResultSetAccess.java:246)
	... 31 more

###############################


#### User summary [took: 0s] [# lines: 19]
Exception occurred when processing report - org.hibernate.HibernateException: Calling method 'createQuery' is not valid without an active transaction (Current status: MARKED_ROLLBACK)
	at org.hibernate.context.internal.ThreadLocalSessionContext$TransactionProtectionWrapper.invoke(ThreadLocalSessionContext.java:339)
	at jdk.proxy2/jdk.proxy2.$Proxy159.createQuery(Unknown Source)
	at org.dspace.core.AbstractHibernateDAO.createQuery(AbstractHibernateDAO.java:147)
	at org.dspace.content.dao.impl.MetadataFieldDAOImpl.findByElement(MetadataFieldDAOImpl.java:130)
	at org.dspace.content.MetadataFieldServiceImpl.findByElement(MetadataFieldServiceImpl.java:107)
	at org.dspace.eperson.EPersonServiceImpl.findAll(EPersonServiceImpl.java:316)
	at org.dspace.eperson.EPersonServiceImpl.findAll(EPersonServiceImpl.java:292)
	at org.dspace.health.UserCheck.run(UserCheck.java:42)
	at org.dspace.health.Check.report(Check.java:31)
	at org.dspace.health.Report.run(Report.java:69)
	at org.dspace.health.Report.main(Report.java:196)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:283)
	at org.dspace.app.launcher.ScriptLauncher.handleScript(ScriptLauncher.java:134)
	at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:99)

###############################


#### Log Analyser Check [took: 3s] [# lines: 3]
====
Exception occurred!
org.dspace.discovery.SearchServiceException: Calling method 'createQuery' is not valid without an active transaction (Current status: MARKED_ROLLBACK)
```

https://dspace-org.slack.com/archives/C3V628QNN/p1722459198020639

**Comment by aseyedia** [comment]
I just realized that I hadn't been updating my `relationship-formats.xml` file correctly. It was being botched. Trying again.

**Comment by aseyedia** [comment]
I am including a table with the desired ER setup just for future reference.

**All relationships are bidirectional.**

| Entity       | Relationship                                           | Related Entity |
|--------------|--------------------------------------------------------|----------------|
| OrgUnit      | isOrgUnitOfPerson / isMemberOfOrgUnit                            | Person         |
| OrgUnit      | isOrgUnitOfConceptSet / isConceptSetOfOrgUnit                | ConceptSet     |
| OrgUnit      | isOrgUnitOfStudy / isStudyOfOrgUnit                   | Study          |
| Person       | isAuthorOfConceptSet / isConceptSetOfAuthor                  | ConceptSet     |
| Person       | isEditorOfConceptSet / isConceptSetOfEditor                  | ConceptSet     |
| Person       | isAuthorOfStudy / isStudyOfAuthor                    | Study          |
| Person       | isEditorOfStudy / isStudyOfEditor                    | Study          |
| ConceptSet   | isConceptSetOfStudy / isStudyOfConceptSet                         | Study          |
| Person       | isDQCheckOfPerson / isDQCheckOfPerson                     | DQCheck        |
| Person       | isAuthorOfDocumentation / isDocumentationOfAuthor               | Documentation  |
| Person       | isEditorOfDocumentation / isDocumentationOfEditor               | Documentation  |
| Person       | isAuthorOfPhenotype / isPhenotypeOfAuthor                   | Phenotype      |
| Person       | isEditorOfPhenotype / isPhenotypeOfEditor                   | Phenotype      |
| Study        | isStudyOfPhenotype / isPhenotypeOfStudy                   | Phenotype      |
| Study        | resultsInConceptSet / isPhenotypeOfConceptSet                  | ConceptSet     |
| ConceptSet   | isConceptSetOfPhenotype / isInformedByConceptSet              | Phenotype      |
| ConceptSet   | isDQCheckOfConceptSet / isConceptSetOfDQCheck          | DQCheck        |
| ConceptSet   | isDocumentationOfConceptSet / isConceptSetOfDocumentation | Documentation |

**Comment by aseyedia** [comment]
The solution was so simple that I don't even want to say what I did.

**Comment by gmu2024** [comment]
I would like to hear how you resolved this, as I am encountering the same problem. Thank you.
Gabriela

**Comment by aseyedia** [comment]
@gmu2024 The case was more that I had over-defined the relationships in `virtual-metadata.xml` (you only need to define them going one way) and that I had entered the labels for *some* of the relationships, but not all. So I had to finish labeling the fields in `en.json5`. 

I also experimented with swapping the `left_type` and `right_type` columns in the PSQL backend, which I absolutely didn't need to do, but I realized that DSpace is capable of handling swapped relationships, and the way it does that is it alternations between the field `relationships.is[Y]Of` and `relationships.is[X]Of.[Y]` - I eventually realized that `relationships.is[Y]Of` was the "correct" orientation even if it didn't make sense because it was from the perspective of Y. 

I'm just going to give you the Entity configurations from a similar (but not the same) project, because I think I configured Entities better the second time I did it. I've attached them to this comment.

[relationship-formats.txt](https://github.com/user-attachments/files/16788084/relationship-formats.txt)
[virtual-metadata.txt](https://github.com/user-attachments/files/16788085/virtual-metadata.txt)
[relationship-types.txt](https://github.com/user-attachments/files/16788086/relationship-types.txt)

They're in `txt` format because GH wouldn't accept them otherwise - just change them to `xml`. I'm on the DSpace Slack if you need more help.

**Comment by gmu2024** [comment]
This is very helpful. Many thanks. Gabriela
